### PR TITLE
Upgrades for mdbook 0.2.0, and link style cleanup.

### PR DESCRIPTION
**Pull request by alercah**
See individual commit messages for description of what I did.

I have not verified all links. `mdbook-linkcheck` does not yet support 0.2.0.

### CODE REVIEW

File: src/items/traits.md [hunk 1]
```diff
@@ -3,9 +3,9 @@
 A _trait_ describes an abstract interface that types can implement. This
 interface consists of [associated items], which come in three varieties:
 
-- [functions](items/associated-items.html#associated-functions-and-methods)
-- [types](items/associated-items.html#associated-types)
-- [constants](items/associated-items.html#associated-constants)
+- [functions]
```
Comment (Havvy) [inline src/items/traits.md]: I feel like this is going to cause confusion as just being functions. How about `[function][associated functions]`?

File: src/items/type-aliases.md [hunk 2]
```diff
@@ -29,7 +27,8 @@ let _: F = E::A;  // OK
...
 ```
 
-[IDENTIFIER]: identifiers.html
-[_Generics_]: items/generics.html
-[_WhereClause_]: items/generics.html#where-clauses
-[_Type_]: types.html
+[type]: ../types.html
+[IDENTIFIER]: ../identifiers.html
+[_Generics_]: generics.html
+[_WhereClause_]: html#where-clauses
```
Comment (Havvy) [inline src/items/type-aliases.md]: Missing the `generics.`.

File: src/items/unions.md [hunk 3]
```diff
@@ -143,9 +143,10 @@ generics, trait implementations, inherent implementations, coherence, pattern
...
-in [RFC 1897 "Unions v1.2"](https://github.com/rust-lang/rfcs/pull/1897).
+in [RFC 1897 "Unions v1.2"].
 
-[IDENTIFIER]: identifiers.html
-[_Generics_]: items/generics.html
-[_WhereClause_]: items/generics.html#where-clauses
-[_StructFields_]: items/structs.html
+[RFC 1897 "Unions v1.2"]: https://github.com/rust-lang/rfcs/pull/1897
+[IDENTIFIER]: ../identifiers.html
+[_Generics_]: html
```
Comment (Havvy) [inline src/items/unions.md]: Missing `generics.` and similar for next two lines.

File: src/expressions.md [hunk 4]
```diff
@@ -233,9 +233,9 @@ Implicit borrows may be taken in the following expressions:
 
 Certain types of expressions can be evaluated at compile time. These are called
 _constant expressions_. Certain places, such as in
-[constants](items/constant-items.html) and [statics](items/static-items.html),
+[constants] and [statics](static variables),
```
Comment (Havvy) [inline src/expressions.md]: Square brackets, not round brackets.

File: src/items.md [hunk 5]
```diff
@@ -36,3 +33,19 @@ as if the item was declared outside the scope &mdash; it is still a static item
 qualified by the name of the enclosing item, or is private to the enclosing
 item (in the case of functions). The grammar specifies the exact locations in
 which sub-item declarations may appear.
+
+[modules]: items/modules.html
+[paths]: paths.html
+[modules]: items/modules.html
```
Comment (Havvy) [inline src/items.md]: modules is duplicated.

File: src/types.md [hunk 6]
```diff
@@ -742,3 +720,30 @@ impl Printable for String {
...
+[visibility modifiers]: visibility-and-privacy.html
+[`enum` item]: items/enumerations.html
+[enum variant expression]: expressions/enum-variant-expr.html
+[coercion]: type-coercions.html
+[`union` item]: items/unions.html
+[function]: items/functions.html
+[structs]: #struct-types
+[enumerations]: #enumerated-types
+[unions]: #union-types
+[type definitions]: ../grammar.html#type-definitions
```
Comment (Havvy) [inline src/types.md]: Do we want to be linking to that grammar still?

**Comment by Havvy** [comment]
Oh, I commented on commits as I checked on each commit. Thus some stuff says out of date even though it's not.

**Comment by Havvy** [comment]
This needs to be rebased pretty badly. I think it'd be easier if the alphabetizing commit was its own PR.

**Comment by alercah** [comment]
Rebased and updated, removed the sorting/organization and just stuck on the version upgrade.

**Comment by alercah** [comment]
@steveklabnik

**Comment by ehuss** [comment]
These links don't look correct to me.  They should all be relative to the page they are on.  For example, `introduction.md` has a bunch of `../` links that should not have `../`.  If you visit the page [here on Github](https://github.com/rust-lang-nursery/reference/blob/198488b45afdd16cd0dd8194bb7d3703d19a5ec1/src/introduction.md), none of the links are working.

There also looks to be some unintended changes to `//` comments in `operator-expr.md`.

**Comment by ehuss** [comment]
Oh, `.travis.yml` will need to be updated, too.

**Comment by steveklabnik** [comment]
Oh shucks, it looks like the upstream PR for multi-mdbook support has not actually landed. I thought it had. Let me take care of that ASAP.

**Comment by steveklabnik** [comment]
Now r+'d https://github.com/rust-lang/rust/pull/57495

**Comment by alercah** [comment]
@ehuss Ah right, I thought I reverted all the ones outside of subdirs since I did them by mistake, but missed `introduction.md`. I fixed some I missed too that linkcheck found.

File: src/items/traits.md (hunk 1 above)
Comment (alercah) [reply src/items/traits.md]: I don't want to make functional changes here.

File: src/expressions/array-expr.md [hunk 7]
```diff
@@ -81,11 +80,12 @@ arr[10];                  // warning: index out of bounds
 The array index expression can be implemented for types other than arrays and slices
 by implementing the [Index] and [IndexMut] traits.
 
+[array]: ../types/array.html
 [IndexMut]: ../std/ops/trait.IndexMut.html
```
Comment (ehuss) [inline src/expressions/array-expr.md]: These two std links need to be fixed up.

File: .travis.yml [hunk 8]
```diff
@@ -4,7 +4,7 @@ rust:
   - nightly
 
 install:
-  - travis_retry curl -Lf https://github.com/rust-lang-nursery/mdBook/releases/download/v0.1.7/mdbook-v0.1.7-x86_64-unknown-linux-gnu.tar.gz | tar -xz --directory=$HOME/.cargo/bin
+  - travis_retry curl -Lf https://github.com/rust-lang-nursery/mdBook/releases/download/v0.2.1/mdbook-v0.2.1-x86_64-unknown-linux-gnu.tar.gz | tar -xz --directory=$HOME/.cargo/bin
```
Comment (ehuss) [inline .travis.yml]: Oh, hmm, this needs to be 0.2.3.  @steveklabnik would it be possible to publish a binary for the latest release on github?

**Comment by steveklabnik** [comment]
So, I have good news, and I have bad news. Good news is, I ran this against upstream's linkchecker. Bad news is that I got:

<details>
```text
reference\expressions\array-expr.html:225: broken link - reference\std\ops\trait.Index.html
reference\expressions\array-expr.html:225: broken link - reference\std\ops\trait.IndexMut.html
reference\print.html:2243: broken link - identifiers.html
reference\print.html:2244: broken link - identifiers.html
reference\print.html:2245: broken link - attributes.html
reference\print.html:2246: broken link - items.html
reference\print.html:2249: broken link - items.html
reference\print.html:2280: broken link - paths.html
reference\print.html:2345: broken link - crates-and-source-files.html
reference\print.html:2347: broken link - crates-and-source-files.html
reference\print.html:2348: broken link - attributes.html
reference\print.html:2353: broken link - conditional-compilation.html
reference\print.html:2354: broken link - attributes.html
reference\print.html:2354: broken link - attributes.html
reference\print.html:2354: broken link - attributes.html
reference\print.html:2362: broken link - identifiers.html
reference\print.html:2364: broken link - identifiers.html
reference\print.html:2367: broken link - identifiers.html
reference\print.html:2376: broken link - identifiers.html
reference\print.html:2400: broken link - crates-and-source-files.html
reference\print.html:2401: broken link - items\modules.html
reference\print.html:2405: broken link - items\use-declarations.html
reference\print.html:2407: broken link - items\use-declarations.html
reference\print.html:2432: broken link - attributes.html
reference\print.html:2440: broken link - paths.html
reference\print.html:2441: broken link - paths.html
reference\print.html:2442: broken link - paths.html
reference\print.html:2442: broken link - identifiers.html

reference\print.html:6145: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.Fn.html
reference\print.html:6145: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnMut.html
reference\print.html:6146: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnOnce.html

reference\print.html:8032: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.Fn.html
reference\print.html:8032: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnMut.html
reference\print.html:8032: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnOnce.html
reference\print.html:8032: broken link - special-types-and-traits.html
reference\print.html:8033: broken link - special-types-and-traits.html
reference\print.html:8033: broken link - special-types-and-traits.html
reference\print.html:8033: broken link - special-types-and-traits.html
reference\print.html:8035: broken link - expressions\closure-expr.html
reference\print.html:8134: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnOnce.html
reference\print.html:8140: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnMut.html
reference\print.html:8144: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.Fn.html
reference\print.html:8148: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.Fn.html
reference\print.html:8148: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\ops\trait.FnMut.html

reference\print.html:7914: broken link - C:\Users\steve\src\jamesrust\build\x86_64-pc-windows-msvc\std\vec\struct.Vec.html

```
</details>

... and more, but I won't paste the same line about print.html over and over again...

I am very confused about `print.html`, because I did https://github.com/rust-lang-nursery/mdBook/pull/866 to fix this.

**Comment by steveklabnik** [comment]
> All of the links to .html in-repo need to be updated to .md. This allows links to work on GitHub, and mdbook doesn't handle the .html links properly in the print page (all links not in the root are currently broken on the print page).

ahhhhhh yes

**Comment by steveklabnik** [comment]
The upstream PR has landed! @alercah let me know when you want me to try the linkchecker again :)

**Comment by alercah** [comment]
Apologies for the delay here, had some issues. Should be good to go now, though I can't get linkcheck to ignore the parent directories so I can't be totally sure.

**Comment by ehuss** [comment]
Hm, something's wrong. I swore I tested the print page on 0.2.3, but there are still problems.  There are two issues:
- Links like `items/unions.md` jump out of the print page.  I would assume the 0.1 behavior should be retained where the print page links to itself.
- Links like `../../std/mem/fn.discriminant.html` are not rebased when used from a subdirectory, and are broken.

I'll see if I can figure out what's wrong in the next few days.

**Comment by ehuss** [comment]
I have posted a proposed fix for the print.html links at https://github.com/rust-lang-nursery/mdBook/pull/871.

As for publishing 0.2.3, it looks like the deploy scripts are broken ([travis](https://travis-ci.com/rust-lang-nursery/mdBook/builds/97850452) and [appveyor](https://ci.appveyor.com/project/rust-lang-libs/mdbook/builds/21723156/job/w0ebv0gtfwes4voq)).  I have posted a partial fix for Travis at https://github.com/rust-lang-nursery/mdBook/pull/872.  The API keys need to be fixed since the transition to travis.com.  @steveklabnik if you have permissions, can you take care of that? Or if not, I can try to bug infra and get things worked out.  Alternatively the releases can be made manually (it's just a compressed copy of the mdbook executable), and uploaded manually on the GitHub releases page.

**Comment by mati865** [comment]
There is mdbook 0.3 so this should be unblocked but there are so many conflicts and new links this PR probably has to be redone from scratch.

**Comment by ehuss** [comment]
@mati865 There was another bug blocking the update of the reference and cargo (https://github.com/rust-lang-nursery/mdBook/pull/963).  It is now fixed, I just need to make a new release.  I'll be posting a new PR here to replace this one when things are ready.  It will require a coordinated update due to some changes to anchor tags.  I think I'll be able to eradicate 0.1 during that update.

**Comment by ehuss** [comment]
I'm going to close this for now.  The reference has been updated to mdbook 0.3 in #641.  If there are any link cleanups we want to do, we can always do them in future PRs. Thanks @alercah, sorry it took so long to iron out all the kinks in mdbook to make this work.

**Pull request closed without merging**